dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/HostSelectionWithFallback.java

```diff
@@ -383,18 +386,22 @@ public void initWithHosts(Map<Host, HostConnectionPool<CL>> hPools) {
         topology.set(createTokenPoolTopology(allHostTokens));
 	}
 
-    /*package private*/ int calculateReplicationFactor(List<HostToken> allHostTokens) {
-        Map<Long, Integer> groups = new HashMap<>();
+	int calculateReplicationFactor(List<HostToken> allHostTokens) {
+		return calculateReplicationFactorForDC(allHostTokens, null);
+	}
 
-        Set<HostToken> uniqueHostTokens = new HashSet<>(allHostTokens);
+	int calculateReplicationFactorForDC(List<HostToken> allHostTokens, String dataCenter) {
+		Map<Long, Integer> groups = new HashMap<>();
 
-        String dataCenter = cpConfig.getLocalDataCenter();
-        if (dataCenter == null) {
-            dataCenter = localRack.substring(0, localRack.length() - 1);
-        }
+		Set<HostToken> uniqueHostTokens = new HashSet<>(allHostTokens);
+		if (dataCenter == null) {
+		    if(localRack != null) {
+				dataCenter = localRack.substring(0, localRack.length() - 1);
+			}
+		}
 
-        for (HostToken hostToken: uniqueHostTokens) {
-        	if (hostToken.getHost().getRack().contains(dataCenter)) {
+		for (HostToken hostToken: uniqueHostTokens) {
+		    if(dataCenter == null || hostToken.getHost().getRack().contains(dataCenter)) {
```

Comment:
This block of code seems like it only wants to obtain the hosts that belong in a specific datacenter. By adding the 'dataCenter == null' check, wouldn't that mean ALL datacenters are now added to the 'groups' map?

If that's correct, then it would probably make more sense to set a default DC as an 'else' condition to the 'if' statement in L398.